<h1 align="center">
    pbccs - Generate Accurate Consensus Sequences from a Single SMRTbell
</h1>

<p align="center">
  <img src="http://www.evolvedmicrobe.com/CCS.png" alt="Image of SMRTbell"/>
</p>

### Input

The ccs program needs a .subreads.bam file containing the subreads for each SMRTbell sequenced.  Older versions of the PacBio RS software outputted data in bas.h5 files, while the new software outputs BAM files.  If you have a bas.h5 file from the older software, you will need to convert it into a BAM.  This can be done with the tool bax2bam which simply needs the name of any bas.h5 files to convert and the prefix of the output file.  Assuming your original file is named mydata.bas.h5, you can produce a file mynewbam.subreads.bam with the following command.

    bax2bam -o mynewbam mydata.1.bax.h5 mydata.2.bax.h5 mydata.3.bax.h5


### Running CCS

CCS can be run at the command line as follows:

    ccs [OPTIONS] INPUT OUTPUT

An example command would be:

    ccs --minLength=100 myData.subreads.bam myResult.bam

The CCS program can be run with a variety of options described below.


|           Option                      |     Example (Defaults)                                |                                                                                                                                                                                                                                                                                                    Explanation                                                                                                                                                                                                                                                                                                     |
| ------------------------------------- | ----------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Input File Name                       | myData.subreads.bam                                   | The name of a single subreads.bam or a subreadset.xml file to be processed. (Example = `myData.subreads.bam`).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| Output File Name                      | myResult.bam                                          | The name of the output BAM file; comes after all other options listed. Valid output files are the BAM and the Dataset .xml formats. (Example = `myResult.bam`).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Version Number                        | --version                                             | Prints the version number.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| Report File Name                      | --reportFile=ccs_report.txt                           | Contains a result tally of the outcomes for all ZMWs that were processed. If no file name is given, the report is output to the file `ccs_report.txt` In addition to the count of successfully-produced consensus sequences, this file lists how many ZMWs failed various data quality filters (SNR too low, not enough full passes, and so on) and is useful for diagnosing unexpected drops in yield. |
| Minimum SNR                           | --minSnr=2.5                                          | Removes data that is likely to contain deletions. SNR is a measure of the strength of signal for all 4 channels (A, C, G, T) used to detect base pair incorporation. This value sets the threshold for minimum required SNR for any of the four channels. Data with SNR < 2.5 is typically considered lower quality. |
| Minimum Read Score                    | ----minReadScore=0.75                                 | Specifies the minimum value for the predicted quality of any subread used for CCS. Note that this filters the input to CCS (the subread quality must be above this value), whereas the `--minPredictedAccuracy` option filters the output (the predicted consensus sequence must be above a certain predicted accuracy).|
| Minimum Length                        | --minLength=10                                        | Specifies the minimum length requirement for the median size insert so that the ZMW is considered for processing, minimum length of aligned insert reads to generate a consensus sequence, and minimum length of the draft consensus to be used for further polishing. If the targeted template is known to be a particular size range, this can filter out alternative DNA templates.                                                                                                                                                                                                                                                                                                                                                                                   |
| Maximum Length                        | --maxLength=21000                                     | Specifies the maximum length for both subreads to process as well as the consensus sequence to generate. For robust results while avoiding unnecessary computation on unusual data, set to ~20% above the largest expected insert size.                                                                                                                                                                                                                                                                                                                                        |
| Minimum Number of Passes              | --minPasses=3                                         | Specifies the minimum number of passes for a ZMW to be emitted. This is the number of full passes. Full passes must have an adapter hit before and after the insert sequence and so do not include any partial passes at the start and end of the sequencing reaction.                                                                                                                                                                                                                                                     |
| Minimum Predicted Accuracy            | --minPredictedAccuracy=0.9                            | Specifies the minimum predicted accuracy of a read. CCS generates an accuracy prediction for each read, defined as the expected percentage of matches in an alignment of the consensus sequence to the true read. A value of 0.99 indicates that only reads expected to be 99% accurate are emitted.                                                                                                                                                                                                                                                                                                                       |
| Maximum Dropped Fraction              | --maxDropFraction=0.34                                | Specifies the maximum number of subreads that can be dropped before the entire ZMW is discarded. Typically, very few reads should be discarded, but if a high proportion are, then the entire ZMW is dropped.                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| ZMWs to Process                       | --zmws=0-100000000000                                 | Specifies the consensus sequence for only a subset of ZMWs. ZMWs can be specified either by range (Example: `--zmws=1-2000`), by values (Example: `--zmws=5,10,20`), or by both (Example: `--zmws=5-10,35,1000-2000`). Use a comma-separated list with no spaces.|
| Number of Threads to Use              | --numThreads=0                                        | Specifies how many threads to use while processing. By default, CCS will use as many threads as there are available cores to minimize processing time, but fewer threads can be specified here.                                                                                                                                                                                                                                                                                                                                                                                                                             |
| Log File                              | --logFile=mylog.txt                                   | The name of a log file to use. If none is given, the logging information is printed to STDERR.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| Log level verbosity                   | --logLevel=INFO                                       | Specifies verbosity of log data to produce. By setting `--logLevel=DEBUG`, you can obtain detailed information on what ZMWs were dropped during processing, as well as any errors which may have appeared.                                                                                                                                                                                                                                                                                                                                                                                                                         |
| Disable Polishing                     | --noPolish                                            | After constructing the initial template, do not proceed with the polishing steps. This is significantly faster, but generates less accurate data with no RQ or QUAL values associated with each base.                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Analyze strands  separately           | --byStrand                                            | Separately generates a consensus sequence from the forward and reverse strands. Useful for identifying heteroduplexes formed during sample preparation.                                                                                                                                                                                                                                                                                                                                                                                                                     |
| Overwrite output file                 | --force                                               | Overwrites the output file if it already exists.                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Filter subreads of poor identity      | --minIdentity=0.82                                    | Applied in the polishing stage, this threshold discards any subread which has alignment identity lower than this value. The identity is calculated by aligning a subread to the draft consensus sequence.                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Add rich QV tags                      | --richQVs                                             | Adds additional tags to the BAM output. The tags contain quality tracks for the deletions (`dq`), insertions (`iq`) and substitutions (`sq`).                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Output tool contract                  | --emit-tool-contract                                  | Outputs the tool contract to `stdout`.                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Specify params via a tool contract    | --resolved-tool-contract=resolved-tool-contract.json  | Instead of specifying arguments manually on the command line, a resolved tool contract can be used. Settings for each of the parameters are loaded from this file.                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Custom model path                     | --modelPath                                           | Specifies the path to a model file or directory containing model files.                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Custom model name                     | --modelSpec                                           | Specifies the name of the chemistry or model to use, overriding the default selection.                                                                                                                                                                                                                                                                                                                                                                                                                        |

### Output

When the ccs program finishes, it outputs a BAM file with one entry for each consensus sequence derived from a ZMW.  BAM is a general file format for storing sequence data, which is described fully by the SAM/BAM working group [here](https://samtools.github.io/hts-specs/SAMv1.pdf).  The CCS output format is a version of this general format, where the consensus sequence is represented by the "Query Sequence" and several tags have been added to provide additional meta information.

An example BAM entry for a consensus as seen by [samtools](http://samtools.sourceforge.net/) is shown below.

```
     m141008_060349_42194_c100704972550000001823137703241586_s1_p0/63/ccs	4	*	0	255	*	*	0	0	CCCGGGGATCCTCTAGAATGC	~~~~~~~~~~~~~~~~~~~~~	RG:Z:83ba013f np:i:35 rq:f:0.999682	sn:B:f,11.3175,6.64119,11.6261,14.5199	t1:f:1.412	t2:f:0.0610001	t3:f:7.338	zm:i:63

 ```

 A description of each of the common separated fields is given below.

|        Name        |                                                                                                                                                                            Definition                                                                                                                                                                                                |
| ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Query Name         | Movie Name / ZMW # / ccs.                                                                                                                                                                                                                                                                                                                                                              |
| FLAG               | Required by format but meaningless in this context.  Always set to indicate the read is unmapped (`4`).                                                                                                                                                                                                                                                                                 |
| Reference Name     | Required by format but meaningless in this context.  Always set to `*`.                                                                                                                                                                                                                                                                                                               |
| Mapping Start      | Required by format but meaningless in this context.  Always set to `0`.                                                                                                                                                                                                                                                                                                               |
| Mapping Quality    | Required by format but meaningless in this context.  Always set to `255`.                                                                                                                                                                                                                                                                                                             |
| CIGAR              | Required by format but meaningless in this context.  Always set to `*`.                                                                                                                                                                                                                                                                                                               |
| RNEXT              | Required by format but meaningless in this context.  Always set to `*`.                                                                                                                                                                                                                                                                                                               |
| PNEXT              | Required by format but meaningless in this context.  Always set to `0`.                                                                                                                                                                                                                                                                                                               |
| TLEN               | Required by format but meaningless in this context.  Always set to `0`.                                                                                                                                                                                                                                                                                                               |
| Consensus Sequence | This is the consensus sequence generated.                                                                                                                                                                                                                                                                                                                                            |
| Quality Values     | This is the per-base parameteric quality metric.  For more information see the "interpretting QUALs" section.                                                                                                                                                                                                                                                                        |
| RG Tag             | This is the read group identifier.                                                                                                                                                                                                                                                                                                                                                   |
| bc Tag             | This is a 2-entry array of upstream-provided barcode calls for this ZMW.                                                                                                                                                                                                                                                                                                             |
| bq Tag             | This is the quality of the barcode call. (Optional, depends on barcoded inputs).                                                                                                                                                                                                                                                                                                      |
| np Tag             | The number of full passes that went into the subread. (Optional, depends on barcoded inputs)                                                                                                                                                                                                                                                                                         |
| rq Tag             | The predicted read quality.                                                                                                                                                                                                                                                                                                                                                          |
| t1 Tag             | The time (in seconds) spent constructing the draft consensus.                                                                                                                                                                                                                                                                                                                                |
| t2 Tag             | The time (in seconds) spent aligning subreads to the draft consensus, prior to polishing.                                                                                                                                                                                                                                                                                                                                |
| t3 Tag             | The time (in seconds) spent polishing the draft consensus, not counting retries.                                                                                                                                                                                                                                                                                                                                |
| zm Tag             | The ZMW hole number.                                                                                                                                                                                                                                                                                                                                                                 |

## Interpretting QUAL values

The QUAL value of a read is a measure of the posterior likelihood of an error at a particular position.  Increasing QUAL values are associated with a decreasing probability of error.  For the case of Indels and homopolymers, it is there is ambiguity as to which QUAL value is associated with the error probability.  Shown below are different types of alignment errors, with a '*' indicating which sequence BP should be associated with the alignment error.

### Mismatch

```
           *
    ccs: ACGTATA
    ref: ACATATA

```

### Deletion

```
            *
    ccs: AC-TATA
    ref: ACATATA

```

### Insertion

```
           *
    ccs: ACGTATA
    ref: AC-TATA

```

### Homopolymer Insertion or Deletion
Indels should always be left-aligned and the error probability is only given for the first base in a homopolymer.

```
           *                    *
    ccs: ACGGGGTATA     ccs: AC-GGGTATA
    ref: AC-GGGTATA     ref: ACGGGGTATA

```



## CCS Yield Report

After ccs runs, a report is generated that specifies the number of ZMWs that successfully produced consensus sequences, as well as a count of how many ZMWs did not produce a consensus sequence for various reasons.  The entries in this report, as well as parameters that can be set to change to increase of decrease the number of ZMWs that pass various filters, are explained in the table below.

| ZMW result                           | Parameter(s) affecting result              | Description                                                                                                                                                                                                                                                                                                   |
|--------------------------------------|--------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Success (without retry)              | All custom processing settings             | The number of CCS reads successfully produced on the first attempt, using the fast windowed approach.                                                                                                                                                                                                         |
| Success (with retry)                 | All custom processing settings             | The number of CCS reads successfully produced on the second attempt, using the slower and more sensitive pipeline. The retry attempt is triggered automatically in case the first attempt fails. Only if the second attempt fails, the ZMW is then classified in one of the Fail categories described below.  |
| Below SNR threshold                  | --minSnr                                   | The ZMW had at least one channel's SNR below the minimum threshold.                                                                                                                                                                                                                                           |
| No usable subreads                   | --minReadScore, --minLength, --maxLength   | The ZMW had no usable subreads. Either there were no subreads, or all the subreads were below the minimum quality threshold, or were above/below the specified length thresholds.                                                                                                                             |
| Insert size too long                 | --maxLength                                | The consensus sequence was above the maximum length threshold. Note that if all the input subreads were already above this threshold, they would all have been excluded, leading to a "No usable subreads" result.                                                                                            |
| Insert size too small                | --minLength                                | The consensus sequence was below the minimum length threshold. Note that if all the input subreads were already below this threshold, they would all have been excluded, leading to a "No usable subreads" result.                                                                                            |
| Not enough full passes               |  --minPasses                               | There were not enough subreads that had an adapter at the start and end of the subread (a "full pass").                                                                                                                                                                                                       |
| Too many unusable subreads           |  --maxDropFraction                         | The ZMW had too many subreads that could not be used.                                                                                                                                                                                                                                                         |
| CCS did not converge                 | None                                       | The consensus sequence did not converge after the maximum number of allowed rounds of polishing.                                                                                                                                                                                                              |
| CCS below minimum predicted accuracy | --minPredictedAccuracy                     | Each CCS read has a predicted level of accuracy associated with it. Reads that are below the minimum specified threshold are removed.                                                                                                                                                                         |
| Unknown error during processing      | None                                       | These should not occur.                                                                                                                                                                                                                                                                                       |

